localstack/packages/__init__.py

```diff
@@ -0,0 +1,349 @@
+        """
+        Returns the installation directory. The directory can differ based on the installation target and version.
+        :return: str representation of the installation directory path or None if the package is not installed anywhere
+        """
+        for target in InstallTarget:
+            directory = self._get_install_dir(target)
+            if directory and os.path.exists(directory):
+                return directory
+
+    def get_executables_path(self) -> str | None:
+        """
+        TODO check if this is useful?
+        Returns the path to the executable of the package (i.e. the binary / script / jar / ... which represents the
+        main executable of the specific package installation).
+        :return: str representation of the executable path or None if the package is not installed anywhere
+        """
+        directory = self.get_installed_dir()
+        if directory:
+            return self._build_executables_path(directory)
+
+    def _get_install_dir(self, target: InstallTarget) -> str:
+        """
+        Builds the installation directory for a specific target.
+        :param target: to create the installation directory path for
+        :return: str representation of the installation directory for the given target
+        """
+        return os.path.join(target.value, self.name, self.version)
+
+    def _build_executables_path(self, install_dir: str) -> str:
+        """
+        Internal function to build the executable path for a specific installation directory.
+        Needs to be implemented by subclasses.
+        :return: str representation of the executable path in the given installation directory
+        """
+        raise NotImplementedError()
+
+    def _prepare_installation(self, target: InstallTarget) -> None:
+        """
+        Internal function to prepare an installation, f.e. by downloading some data or installing an OS package repo.
+        Can be implemented by specific installers.
+        :param target: of the installation
+        :return: None
+        """
+        pass
+
+    def _install(self, target: InstallTarget) -> None:
+        """
+        Internal function to perform the actual installation.
+        Must be implemented by specific installers.
+        :param target: of the installation
+        :return: None
+        """
+        raise NotImplementedError()
+
+    def _post_process(self, target: InstallTarget) -> None:
+        """
+        Internal function to perform some post-processing, f.e. patching an installation or creating symlinks.
+        :param target: of the installation
+        :return: None
+        """
+        pass
+
+
+# Lock which is used for OS package installations (to avoid locking issues)
+OS_PACKAGE_INSTALL_LOCK = threading.RLock()
+# Cache directory for APT / the debian package manager.
+_DEBIAN_CACHE_DIR = os.path.join(config.dirs.cache, "apt")
+
+
+class OSPackageInstaller(PackageInstaller, ABC):
```

Comment:
I do agree that this is a bit clunky, but the reasoning here was that we will not introduce other operating systems anytime soon and the compositional approach used for the `OSSpecificPackageInstaller` felt then over-engineered.
We only have two versions here because there is a set of users which need to build their images on their own on RedHat based images.
I also wanted to basically stress that there _needs_ to be an implementation for these two operating systems if it's in any way possible (and make it "hard" not to).
However, this is totally debatable. It would be very simple to migrate the `OSSpecificPackageInstaller` into the new package abstraction.

Reply:
I think that's fair for now. We can always change it if we find it's a problem. Ideally we don't need os-level packages in the first place